wgengine/magicsock/magicsock.go

```diff
@@ -3186,6 +3292,37 @@ func (de *discoEndpoint) noteActiveLocked() {
 	}
 }
 
+// cliPing starts a ping for the "tailscale ping" command. res is value to call cb with,
+// already partially filled.
+func (de *discoEndpoint) cliPing(res *ipnstate.PingResult, cb func(*ipnstate.PingResult)) {
+	now := time.Now()
+
+	de.mu.Lock()
+	udpAddr, derpAddr := de.addrForSendLocked(now)
+	// Session active? If so, avoid sending a ping. They're already going out every
+	// 2 seconds. Just return what we know.
+	if !udpAddr.IsZero() && now.Before(de.trustBestAddrUntil) {
+		if es, ok := de.endpointState[udpAddr]; ok && !es.lastPing.IsZero() {
+			pr := es.recentPongs[es.recentPong]
+			res.Endpoint = udpAddr.String()
+			res.LatencySeconds = pr.latency.Seconds()
+			de.mu.Unlock()
+			cb(res)
+			return
+		}
+	}
+
+	de.pendingCliPingRes = res
+	de.pendingCliPingFunc = cb
+	de.sendPingsLocked(now, true)
```

Comment:
Wait, this is sending multiple pings? The API from the CLI side implies a single ping with a single response.

Reply:
"tailscale ping" brings up a new connection to a peer if it's idle, which implies all the normal path discovery.